File: src/extract.ts
import { join, isAbsolute } from 'node:path';
import { spawn } from 'node:child_process';
import { rm, cp, chmod, readdir } from 'node:fs/promises';
import fg from 'fast-glob';
import { ensureDir, pathExists } from './utils/fs.js';
import { log } from './utils/log.js';
import type { Channel } from './types/index.js';

export interface ExtractOptions {
  channel: Channel;
  version: string;
  dataRoot: string;
  p4kPath?: string;
  unp4k?: {
    bin: string;
    args?: string[];
  };
  enableUnp4k?: boolean;
  forceUnp4k?: boolean;
  unforge?: {
    bin: string;
    args?: string[];
  };
  enableUnforge?: boolean;
  scDataDumper?: {
    bin: string;
    args?: string[];
    outputDir?: string;
  };
  enableScDataDumper?: boolean;
  wineBin?: string;
}

export interface ExtractResult {
  rawDir: string;
  normalizedDir: string;
  discoveredFiles: string[];
}

const REQUIRED_FILES = ['manufacturers.json', 'ships.json', 'items.json'];
const OPTIONAL_FILES = [
  'ship_variants.json',
  'hardpoints.json',
  'item_stats.json',
  'ship_stats.json',
  'installed_items.json'
];

export async function extract(opts: ExtractOptions): Promise<ExtractResult> {
  const rawDir = join(opts.dataRoot, 'raw', opts.channel, opts.version);
  const normalizedDir = join(opts.dataRoot, 'normalized', opts.channel, opts.version);

  const rawDirHasData = await directoryHasPayload(rawDir);

  const shouldRunUnp4k = Boolean(
    opts.enableUnp4k &&
      opts.p4kPath &&
      opts.unp4k &&
      (opts.forceUnp4k || !rawDirHasData)
  );

  if (shouldRunUnp4k) {
    await rm(rawDir, { recursive: true, force: true });
    await ensureDir(rawDir);

    const argSets = opts.unp4k!.args && opts.unp4k!.args.length
      ? [opts.unp4k!.args]
      : [
          ['{{p4k}}', '*.xml'],
          ['{{p4k}}', '*.ini']
        ];

    for (const args of argSets) {
      await runUnp4k({
        bin: opts.unp4k!.bin,
        args,
        p4kPath: opts.p4kPath!,
        outputDir: rawDir,
        wineBin: opts.wineBin
      });
    }
  }

  const shouldRunUnforge = Boolean(opts.enableUnforge && opts.unforge);
  if (shouldRunUnforge) {
    await runUnforge({
      bin: opts.unforge!.bin,
      args: opts.unforge!.args ?? [],
      targetDir: rawDir,
      wineBin: opts.wineBin
    });
  }

  const shouldRunScDataDumper = Boolean(opts.enableScDataDumper && opts.scDataDumper);
  if (shouldRunScDataDumper) {
    await runScDataDumper({
      bin: opts.scDataDumper!.bin,
      args: opts.scDataDumper!.args ?? [],
      inputDir: rawDir,
      outputDir: opts.scDataDumper!.outputDir ?? rawDir,
      wineBin: opts.wineBin
    });
  }

  if (!(await pathExists(rawDir))) {
    if (opts.p4kPath) {
      throw new Error(
        `Raw data directory missing after extraction steps. Expected files under ${rawDir}.`
      );
    }
    throw new Error(`Raw data directory missing: ${rawDir}`);
  }

  const missing: string[] = [];
  for (const file of REQUIRED_FILES) {
    if (!(await pathExists(join(rawDir, file)))) missing.push(file);
  }
  if (missing.length) {
    throw new Error(`Missing required raw data files: ${missing.join(', ')}`);
  }

  const optionalMissing: string[] = [];
  for (const file of OPTIONAL_FILES) {
    if (!(await pathExists(join(rawDir, file)))) optionalMissing.push(file);
  }
  if (optionalMissing.length) {
    log.warn('Optional raw files missing (will be skipped):', optionalMissing);
  }

  const discoveredFiles = await fg(['**/*.json'], { cwd: rawDir });

  await ensureDir(normalizedDir);
  log.info('Extraction check complete', { rawDir, normalizedDir, files: discoveredFiles.length });

  return { rawDir, normalizedDir, discoveredFiles };
}

interface Unp4kRun {
  bin: string;
  args: string[];
  p4kPath: string;
  outputDir: string;
  wineBin?: string;
}

async function runUnp4k(config: Unp4kRun) {
  const templateArgs = config.args.length ? config.args : ['{{p4k}}', '*.xml'];
  const absoluteP4k = isAbsolute(config.p4kPath)
    ? config.p4kPath
    : join(process.cwd(), config.p4kPath);

  let hasP4k = false;

  const resolvedArgs = templateArgs.map((arg) => {
    if (arg === '{{p4k}}') {
      hasP4k = true;
      return absoluteP4k;
    }
    if (arg === '{{output}}') {
      return config.outputDir;
    }

    const replaced = arg
      .replace('{{p4k}}', absoluteP4k)
      .replace('{{output}}', config.outputDir);

    if (replaced.includes(absoluteP4k)) hasP4k = true;
    return replaced;
  });

  if (!hasP4k) {
    resolvedArgs.unshift(absoluteP4k);
  }
  await ensureDir(config.outputDir);

  log.info('Running unp4k', {
    bin: config.bin,
    args: resolvedArgs,
    p4k: absoluteP4k,
    output: config.outputDir
  });

  await executeExternal(config.bin, resolvedArgs, {
    wineBin: config.wineBin,
    cwd: config.outputDir
  });
}

interface UnforgeRun {
  bin: string;
  args: string[];
  targetDir: string;
  wineBin?: string;
}

async function runUnforge(config: UnforgeRun) {
  // ASSUMPTION: unforge wird im extrahierten Verzeichnis ausgeführt und akzeptiert einen
  // Pfad als Argument. Standardmäßig nutzen wir "{{input}}" als Platzhalter.
  const templateArgs = config.args.length ? config.args : ['{{input}}'];
  const absoluteTargetDir = isAbsolute(config.targetDir)
    ? config.targetDir
    : join(process.cwd(), config.targetDir);
  let hasInput = false;

  const resolvedArgs = templateArgs.map((arg) => {
    if (arg === '{{input}}') {
      hasInput = true;
      return absoluteTargetDir;
    }

    const replaced = arg.replace('{{input}}', absoluteTargetDir);
    if (replaced.includes(absoluteTargetDir)) hasInput = true;
    return replaced;
  });

  if (!hasInput) {
    throw new Error('unforge arguments must include the {{input}} placeholder.');
  }

  log.info('Running unforge', {
    bin: config.bin,
    args: resolvedArgs,
    target: absoluteTargetDir
  });

  await executeExternal(config.bin, resolvedArgs, {
    wineBin: config.wineBin,
    cwd: absoluteTargetDir
  });
}

interface ScDataDumperRun {
  bin: string;
  args: string[];
  inputDir: string;
  outputDir: string;
  wineBin?: string;
}

async function runScDataDumper(config: ScDataDumperRun) {
  if (!config.args.length) {
    log.warn('scdatadumper configured without arguments – skipping converter step.');
    return;
  }

  const templateArgs = config.args;
  const placeholderTypes: Array<'input' | 'output' | null> = [];
  let hasInput = false;
  let hasOutput = false;
  const absoluteInputDir = isAbsolute(config.inputDir)
    ? config.inputDir
    : join(process.cwd(), config.inputDir);
  const absoluteOutputDir = isAbsolute(config.outputDir)
    ? config.outputDir
    : join(process.cwd(), config.outputDir);

  const resolvedArgs = templateArgs.map((arg) => {
    if (arg === '{{input}}') {
      hasInput = true;
      placeholderTypes.push('input');
      return absoluteInputDir;
    }
    if (arg === '{{output}}') {
      hasOutput = true;
      placeholderTypes.push('output');
      return absoluteOutputDir;
    }

    let replaced = arg;
    let type: 'input' | 'output' | null = null;

    if (arg.includes('{{input}}')) {
      replaced = replaced.replace(/{{input}}/g, absoluteInputDir);
      hasInput = true;
      type = 'input';
    }
    if (arg.includes('{{output}}')) {
      replaced = replaced.replace(/{{output}}/g, absoluteOutputDir);
      hasOutput = true;
      type = 'output';
    }

    if (type === null) {
      if (replaced.includes(absoluteOutputDir)) {
        hasOutput = true;
        type = 'output';
      } else if (replaced.includes(absoluteInputDir)) {
        hasInput = true;
        type = 'input';
      }
    }

    placeholderTypes.push(type);
    return replaced;
  });

  if (!hasInput || !hasOutput) {
    throw new Error('scdatadumper arguments must include {{input}} and {{output}} placeholders.');
  }

  const normalizedBin = config.bin.split(/[/\\]/).pop()?.toLowerCase() ?? '';
  const composePath = join(process.cwd(), 'bins', 'scdatadumper', 'compose.yaml');
  const isComposeWrapper = normalizedBin === 'docker-compose' || normalizedBin === 'podman-compose';
  const isComposeCli = normalizedBin === 'docker' || normalizedBin === 'podman';
  const argsIncludeExec = resolvedArgs.includes('exec');

  const hostBaseDir = join(process.cwd(), 'bins', 'scdatadumper');
  const hostImportDir = join(hostBaseDir, 'import');
  const hostExportDir = join(hostBaseDir, 'export');
  const cliEntrypoint = join(hostBaseDir, 'cli.php');

  await rm(hostImportDir, { recursive: true, force: true });
  await cp(absoluteInputDir, hostImportDir, { recursive: true, force: true });
  await chmod(hostImportDir, 0o777);

  await rm(hostExportDir, { recursive: true, force: true });
  await ensureDir(hostExportDir);
  await chmod(hostExportDir, 0o777);

  const containerInputDir = '/var/www/html/import';
  const containerOutputDir = '/var/www/html/export';

  const finalArgs = resolvedArgs.map((arg, idx) => {
    if (isComposeWrapper || isComposeCli) {
      const type = placeholderTypes[idx];
      if (type === 'input') {
        return containerInputDir;
      }
      if (type === 'output') {
        return containerOutputDir;
      }
      return arg
        .replace(absoluteInputDir, containerInputDir)
        .replace(absoluteOutputDir, containerOutputDir)
        .replace(config.inputDir, containerInputDir)
        .replace(config.outputDir, containerOutputDir);
    }
    return arg;
  });

  log.info('Running scdatadumper', {
    bin: config.bin,
    args: finalArgs,
    input: config.inputDir,
    output: config.outputDir
  });

  if (argsIncludeExec && (isComposeWrapper || isComposeCli)) {
    const upArgs = isComposeWrapper
      ? ['-f', composePath, 'up', '-d']
      : ['compose', '-f', composePath, 'up', '-d'];
    await executeExternal(config.bin, upArgs, { wineBin: config.wineBin });
  }
  const needsGenerateCache = true;

  if (isComposeWrapper || isComposeCli) {
    const execPrefix = isComposeWrapper
      ? ['-f', composePath, 'exec', 'scdatadumper']
      : ['compose', '-f', composePath, 'exec', 'scdatadumper'];

    if (needsGenerateCache) {
      const generateArgs = [...execPrefix, 'php', 'cli.php', 'generate:cache', containerInputDir];
      await executeExternal(config.bin, generateArgs, { wineBin: config.wineBin });
    }

    await executeExternal(config.bin, finalArgs, { wineBin: config.wineBin });
  } else {
    if (normalizedBin === 'php' || normalizedBin.endsWith('.php')) {
      const resolvePhpArgs = (args: string[]) =>
        args.map((arg) => {
          if (arg === 'cli.php') return cliEntrypoint;
          if (arg === config.inputDir) return absoluteInputDir;
          if (arg === config.outputDir) return absoluteOutputDir;
          if (arg === absoluteInputDir) return absoluteInputDir;
          if (arg === absoluteOutputDir) return absoluteOutputDir;
          return arg;
        });

      if (needsGenerateCache) {
        await executeExternal(
          config.bin,
          resolvePhpArgs(['cli.php', 'generate:cache', absoluteInputDir]),
          { wineBin: config.wineBin, cwd: hostBaseDir }
        );
      }
      await executeExternal(config.bin, resolvePhpArgs(finalArgs), {
        wineBin: config.wineBin,
        cwd: hostBaseDir
      });
    } else {
      await executeExternal(config.bin, finalArgs, { wineBin: config.wineBin });
    }
  }

  if (await pathExists(hostExportDir)) {
    await ensureDir(absoluteOutputDir);
    await cp(hostExportDir, absoluteOutputDir, { recursive: true, force: true });
  }
}

async function executeExternal(
  bin: string,
  args: string[],
  options: { wineBin?: string; cwd?: string } = {}
) {
  const isPathLike = bin.includes('/') || bin.includes('\\');
  const absoluteBin = isPathLike ? (isAbsolute(bin) ? bin : join(process.cwd(), bin)) : bin;
  const needsWine =
    process.platform !== 'win32' && absoluteBin.toLowerCase().endsWith('.exe');
  const spawnBin = needsWine ? options.wineBin ?? 'wine' : absoluteBin;
  const spawnArgs = needsWine ? [absoluteBin, ...args] : args;

  await new Promise<void>((resolve, reject) => {
    const child = spawn(spawnBin, spawnArgs, {
      stdio: 'inherit',
      cwd: options.cwd ?? process.cwd()
    });
    child.on('error', reject);
    child.on('exit', (code) => {
      if (code === 0) {
        resolve();
      } else {
        reject(new Error(`${spawnBin} exited with code ${code}`));
      }
    });
  });
}

async function directoryHasPayload(path: string): Promise<boolean> {
  if (!(await pathExists(path))) return false;
  const entries = await readdir(path);
  return entries.some((entry) => entry !== '.gitkeep');
}

File: internal/extract/extract.go
package extract

import (
	"context"
	"errors"
	"fmt"
	"os"
	"path/filepath"
	"runtime"
	"sort"
	"strings"

	"github.com/ArisCorporation/sc-goetl/internal/config"
	"github.com/ArisCorporation/sc-goetl/internal/utils"
)

// Result captures the outcome of the extract phase.
type Result struct {
	RawDir        string
	NormalizedDir string
	Discovered    []string
}

var (
	requiredFiles = []string{"manufacturers.json", "ships.json", "items.json"}
	optionalFiles = []string{
		"ship_variants.json",
		"hardpoints.json",
		"item_stats.json",
		"ship_stats.json",
		"installed_items.json",
	}
)

// Run executes the extract phase of the ETL, invoking external tooling as configured.
func Run(ctx context.Context, cfg *config.Config) (*Result, error) {
	rawDir := filepath.Join(cfg.DataRoot, "raw", cfg.Channel, cfg.Version)
	normalizedDir := filepath.Join(cfg.DataRoot, "normalized", cfg.Channel, cfg.Version)

	rawDirHasData, err := utils.DirectoryHasPayload(rawDir)
	if err != nil && !errors.Is(err, os.ErrNotExist) {
		return nil, err
	}

	shouldRunUnp4k := cfg.Unp4kEnable &&
		cfg.P4KPath != "" &&
		cfg.Unp4k.Bin != "" &&
		(cfg.ForceUnp4k || !rawDirHasData)

	if shouldRunUnp4k {
		if err := os.RemoveAll(rawDir); err != nil {
			return nil, fmt.Errorf("remove raw directory: %w", err)
		}
		if err := utils.EnsureDir(rawDir); err != nil {
			return nil, fmt.Errorf("ensure raw directory: %w", err)
		}

		argSets := [][]string{}
		if len(cfg.Unp4k.Args) > 0 {
			argSets = append(argSets, cfg.Unp4k.Args)
		} else {
			argSets = append(argSets,
				[]string{"{{p4k}}", "*.xml"},
				[]string{"{{p4k}}", "*.ini"},
			)
		}
		for _, args := range argSets {
			if err := runUnp4k(ctx, runUnp4kConfig{
				Bin:       cfg.Unp4k.Bin,
				Args:      args,
				P4KPath:   cfg.P4KPath,
				OutputDir: rawDir,
				WineBin:   cfg.WineBin,
			}); err != nil {
				return nil, err
			}
		}
	}

	if cfg.UnforgeEnable && cfg.Unforge.Bin != "" {
		if err := runUnforge(ctx, runUnforgeConfig{
			Bin:       cfg.Unforge.Bin,
			Args:      cfg.Unforge.Args,
			TargetDir: rawDir,
			WineBin:   cfg.WineBin,
		}); err != nil {
			return nil, err
		}
	}

	if cfg.ScdEnable && cfg.ScDataDump.Bin != "" {
		if err := runScDataDumper(ctx, runScDataDumperConfig{
			Bin:       cfg.ScDataDump.Bin,
			Args:      cfg.ScDataDump.Args,
			InputDir:  rawDir,
			OutputDir: coalesce(cfg.ScDataDump.OutputDir, rawDir),
			WineBin:   cfg.WineBin,
		}); err != nil {
			return nil, err
		}
	}

	exists, err := utils.PathExists(rawDir)
	if err != nil {
		return nil, err
	}
	if !exists {
		if cfg.P4KPath != "" {
			return nil, fmt.Errorf("raw data directory missing after extraction. expected files under %s", rawDir)
		}
		return nil, fmt.Errorf("raw data directory missing: %s", rawDir)
	}

	if err := validateRequiredFiles(rawDir); err != nil {
		return nil, err
	}

	discovered, err := discoverJSON(rawDir)
	if err != nil {
		return nil, err
	}

	if err := utils.EnsureDir(normalizedDir); err != nil {
		return nil, err
	}

	utils.Logger().Info("Extraction check complete",
		"raw_dir", rawDir,
		"normalized_dir", normalizedDir,
		"files", len(discovered),
	)

	return &Result{
		RawDir:        rawDir,
		NormalizedDir: normalizedDir,
		Discovered:    discovered,
	}, nil
}

type runUnp4kConfig struct {
	Bin       string
	Args      []string
	P4KPath   string
	OutputDir string
	WineBin   string
}

func runUnp4k(ctx context.Context, cfg runUnp4kConfig) error {
	bin := toAbs(cfg.Bin)
	templateArgs := cfg.Args
	if len(templateArgs) == 0 {
		templateArgs = []string{"{{p4k}}", "*.xml"}
	}

	absP4k := cfg.P4KPath
	if !filepath.IsAbs(absP4k) {
		if wd, err := os.Getwd(); err == nil {
			absP4k = filepath.Join(wd, absP4k)
		}
	}

	resolved := []string{}
	hasP4k := false
	for _, arg := range templateArgs {
		switch arg {
		case "{{p4k}}":
			resolved = append(resolved, absP4k)
			hasP4k = true
		case "{{output}}":
			resolved = append(resolved, cfg.OutputDir)
		default:
			value := strings.ReplaceAll(arg, "{{p4k}}", absP4k)
			value = strings.ReplaceAll(value, "{{output}}", cfg.OutputDir)
			if strings.Contains(value, absP4k) {
				hasP4k = true
			}
			resolved = append(resolved, value)
		}
	}
	if !hasP4k {
		resolved = append([]string{absP4k}, resolved...)
	}

	if err := utils.EnsureDir(cfg.OutputDir); err != nil {
		return err
	}

	utils.Logger().Info("Running unp4k", "bin", bin, "args", resolved, "p4k", absP4k, "output", cfg.OutputDir)
	return utils.RunExternal(ctx, bin, resolved, utils.RunOptions{
		WorkingDir: cfg.OutputDir,
		WineBin:    cfg.WineBin,
	})
}

type runUnforgeConfig struct {
	Bin       string
	Args      []string
	TargetDir string
	WineBin   string
}

func runUnforge(ctx context.Context, cfg runUnforgeConfig) error {
	bin := toAbs(cfg.Bin)
	templateArgs := cfg.Args
	if len(templateArgs) == 0 {
		templateArgs = []string{"{{input}}"}
	}

	absTarget := cfg.TargetDir
	if !filepath.IsAbs(absTarget) {
		if wd, err := os.Getwd(); err == nil {
			absTarget = filepath.Join(wd, absTarget)
		}
	}

	resolved := []string{}
	hasInput := false
	for _, arg := range templateArgs {
		switch arg {
		case "{{input}}":
			resolved = append(resolved, absTarget)
			hasInput = true
		default:
			value := strings.ReplaceAll(arg, "{{input}}", absTarget)
			if strings.Contains(value, absTarget) {
				hasInput = true
			}
			resolved = append(resolved, value)
		}
	}
	if !hasInput {
		return errors.New("unforge arguments must include the {{input}} placeholder")
	}

	utils.Logger().Info("Running unforge", "bin", bin, "args", resolved, "target", absTarget)
	return utils.RunExternal(ctx, bin, resolved, utils.RunOptions{
		WorkingDir: absTarget,
		WineBin:    cfg.WineBin,
	})
}

type runScDataDumperConfig struct {
	Bin       string
	Args      []string
	InputDir  string
	OutputDir string
	WineBin   string
}

func runScDataDumper(ctx context.Context, cfg runScDataDumperConfig) error {
	if len(cfg.Args) == 0 {
		utils.Logger().Warn("scdatadumper configured without arguments – skipping converter step.")
		return nil
	}

	bin := toAbs(cfg.Bin)
	absInput := toAbs(cfg.InputDir)
	absOutput := toAbs(cfg.OutputDir)

	argPlaceholders := make([]*string, len(cfg.Args))
	resolved := make([]string, len(cfg.Args))
	for i, arg := range cfg.Args {
		switch arg {
		case "{{input}}":
			resolved[i] = absInput
			argPlaceholders[i] = &cfg.InputDir
		case "{{output}}":
			resolved[i] = absOutput
			argPlaceholders[i] = &cfg.OutputDir
		default:
			value := strings.ReplaceAll(arg, "{{input}}", absInput)
			value = strings.ReplaceAll(value, "{{output}}", absOutput)
			resolved[i] = value
		}
	}

	normalizedBin := strings.ToLower(filepath.Base(bin))
	hostBaseDir := toAbs("bins/scdatadumper")
	hostImportDir := filepath.Join(hostBaseDir, "import")
	hostExportDir := filepath.Join(hostBaseDir, "export")
	cliEntrypoint := filepath.Join(hostBaseDir, "cli.php")

	if err := os.RemoveAll(hostImportDir); err != nil {
		return err
	}
	if err := utils.CopyDir(absInput, hostImportDir); err != nil {
		return err
	}
	if err := os.Chmod(hostImportDir, 0o777); err != nil {
		return err
	}

	if err := os.RemoveAll(hostExportDir); err != nil {
		return err
	}
	if err := utils.EnsureDir(hostExportDir); err != nil {
		return err
	}
	if err := os.Chmod(hostExportDir, 0o777); err != nil {
		return err
	}

	containerInput := "/var/www/html/import"
	containerOutput := "/var/www/html/export"
	isComposeWrapper := normalizedBin == "docker-compose" || normalizedBin == "podman-compose"
	isComposeCLI := normalizedBin == "docker" || normalizedBin == "podman"

	if isComposeCLI {
		if len(resolved) == 0 || resolved[0] != "compose" {
			resolved = append([]string{"compose"}, resolved...)
		}
	}
	if isComposeCLI || isComposeWrapper {
		hasFileFlag := false
		for _, arg := range resolved {
			if arg == "-f" || arg == "--file" {
				hasFileFlag = true
				break
			}
		}
		if !hasFileFlag {
			if isComposeCLI {
				resolved = append([]string{"compose", "-f", filepath.Join(hostBaseDir, "compose.yaml")}, resolved[1:]...)
			} else {
				resolved = append([]string{"-f", filepath.Join(hostBaseDir, "compose.yaml")}, resolved...)
			}
		}
		for i, placeholder := range argPlaceholders {
			if placeholder == nil {
				continue
			}
			switch *placeholder {
			case cfg.InputDir:
				resolved[i] = containerInput
			case cfg.OutputDir:
				resolved[i] = containerOutput
			}
		}
		for i, value := range resolved {
			value = strings.ReplaceAll(value, absInput, containerInput)
			value = strings.ReplaceAll(value, absOutput, containerOutput)
			resolved[i] = value
		}
	}

	utils.Logger().Info("Running scdatadumper", "bin", bin, "args", resolved, "input", cfg.InputDir, "output", cfg.OutputDir)

	if isComposeCLI || isComposeWrapper {
		upArgs := []string{"up", "-d"}
		baseArgs := []string{}
		if isComposeCLI {
			baseArgs = append([]string{"compose", "-f", filepath.Join(hostBaseDir, "compose.yaml")}, upArgs...)
		} else {
			baseArgs = append([]string{"-f", filepath.Join(hostBaseDir, "compose.yaml")}, upArgs...)
		}
		if err := utils.RunExternal(ctx, bin, baseArgs, utils.RunOptions{WineBin: cfg.WineBin}); err != nil {
			return err
		}
		execPrefix := []string{}
		if isComposeCLI {
			execPrefix = []string{"compose", "-f", filepath.Join(hostBaseDir, "compose.yaml"), "exec", "scdatadumper"}
		} else {
			execPrefix = []string{"-f", filepath.Join(hostBaseDir, "compose.yaml"), "exec", "scdatadumper"}
		}
		generateArgs := append(execPrefix, "php", "cli.php", "generate:cache", containerInput)
		if err := utils.RunExternal(ctx, bin, generateArgs, utils.RunOptions{WineBin: cfg.WineBin}); err != nil {
			return err
		}
		if err := utils.RunExternal(ctx, bin, resolved, utils.RunOptions{WineBin: cfg.WineBin}); err != nil {
			return err
		}
	} else {
		if normalizedBin == "php" || strings.HasSuffix(normalizedBin, ".php") {
			resolveArgs := func(args []string) []string {
				out := make([]string, len(args))
				for i, arg := range args {
					switch arg {
					case "cli.php":
						out[i] = cliEntrypoint
					case cfg.InputDir, absInput:
						out[i] = absInput
					case cfg.OutputDir, absOutput:
						out[i] = absOutput
					default:
						value := strings.ReplaceAll(arg, cfg.InputDir, absInput)
						value = strings.ReplaceAll(value, cfg.OutputDir, absOutput)
						out[i] = value
					}
				}
				return out
			}
			if err := utils.RunExternal(ctx, bin, resolveArgs([]string{"cli.php", "generate:cache", absInput}), utils.RunOptions{
				WineBin:    cfg.WineBin,
				WorkingDir: hostBaseDir,
			}); err != nil {
				return err
			}
			if err := utils.RunExternal(ctx, bin, resolveArgs(resolved), utils.RunOptions{
				WineBin:    cfg.WineBin,
				WorkingDir: hostBaseDir,
			}); err != nil {
				return err
			}
		} else {
			if err := utils.RunExternal(ctx, bin, resolved, utils.RunOptions{WineBin: cfg.WineBin}); err != nil {
				return err
			}
		}
	}

	if exists, err := utils.PathExists(hostExportDir); err == nil && exists {
		if err := utils.EnsureDir(absOutput); err != nil {
			return err
		}
		if err := utils.CopyDir(hostExportDir, absOutput); err != nil {
			return err
		}
	}
	return nil
}

func coalesce(values ...string) string {
	for _, value := range values {
		if strings.TrimSpace(value) != "" {
			return value
		}
	}
	return ""
}

func toAbs(path string) string {
	if filepath.IsAbs(path) {
		return path
	}
	if path == "" {
		return ""
	}
	if wd, err := os.Getwd(); err == nil {
		return filepath.Join(wd, path)
	}
	return filepath.Clean(path)
}

func validateRequiredFiles(rawDir string) error {
	for _, file := range requiredFiles {
		exists, err := utils.PathExists(filepath.Join(rawDir, file))
		if err != nil {
			return err
		}
		if !exists {
			return fmt.Errorf("missing required raw data file: %s", file)
		}
	}
	var missing []string
	for _, file := range optionalFiles {
		exists, err := utils.PathExists(filepath.Join(rawDir, file))
		if err != nil {
			return err
		}
		if !exists {
			missing = append(missing, file)
		}
	}
	if len(missing) > 0 {
		utils.Logger().Warn("Optional raw files missing (will be skipped)", "files", missing)
	}
	return nil
}

func discoverJSON(root string) ([]string, error) {
	discovered := []string{}
	err := filepath.WalkDir(root, func(path string, d os.DirEntry, err error) error {
		if err != nil {
			return err
		}
		if d.IsDir() {
			return nil
		}
		if strings.HasSuffix(strings.ToLower(d.Name()), ".json") {
			rel, err := filepath.Rel(root, path)
			if err != nil {
				return err
			}
			discovered = append(discovered, filepath.ToSlash(rel))
		}
		return nil
	})
	if err != nil {
		return nil, err
	}
	sort.Strings(discovered)
	return discovered, nil
}

func init() {
	// Windows path normalisation ensures consistent logging.
	if runtime.GOOS == "windows" {
		for i, file := range requiredFiles {
			requiredFiles[i] = filepath.ToSlash(file)
		}
		for i, file := range optionalFiles {
			optionalFiles[i] = filepath.ToSlash(file)
		}
	}
}
